server/s3_delete_object_action.cc

```diff
@@ -125,6 +134,91 @@ void S3DeleteObjectAction::fetch_object_info_failed() {
   s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
 }
 
+void S3DeleteObjectAction::delete_handler() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  std::string bucket_versioning =
+      bucket_metadata->get_bucket_versioning_status();
+
+  if (bucket_versioning == "Enabled" &&
+      !request->has_query_param_key("versionId")) {
+    create_delete_marker();
+    next();
+  } else {
+    populate_probable_dead_oid_list();
+  }
+
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3DeleteObjectAction::metadata_handler() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  std::string bucket_versioning =
+      bucket_metadata->get_bucket_versioning_status();
+  if (bucket_versioning == "Enabled" &&
+      !request->has_query_param_key("versionId")) {
+    save_delete_marker();
+  } else {
+    delete_metadata();
+  }
+
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3DeleteObjectAction::create_delete_marker() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  // create metadata
+  delete_marker_metadata =
+      object_metadata_factory->create_delete_marker_metadata_obj(
+          request, bucket_metadata->get_object_list_index_layout(),
+          bucket_metadata->get_objects_version_list_index_layout());
+
+  // Generate a version id for the new object.
+  delete_marker_metadata->regenerate_version_id();
+  delete_marker_metadata->reset_date_time_to_current();
+
+  // TODO: update null version
+
+  s3_log(S3_LOG_DEBUG, request_id, "Add delete marker4\n");
+  for (auto it : request->get_in_headers_copy()) {
+    if (it.first.find("x-amz-meta-") != std::string::npos) {
+      s3_log(S3_LOG_DEBUG, request_id,
+             "Writing user metadata on object: [%s] -> [%s]\n",
+             it.first.c_str(), it.second.c_str());
+      delete_marker_metadata->add_user_defined_attribute(it.first, it.second);
+    }
+  }
```

Comment:
I don't think I understand what's going on here. Can users set metadata on delete markers?

Reply:
I created interface in s3_factory to provide delete-marker metadata
```
  delete_marker_metadata =
      object_metadata_factory->create_delete_marker_metadata_obj(
          request, bucket_metadata->get_object_list_index_layout(),
          bucket_metadata->get_objects_version_list_index_layout());
```
which create metadata object and set delete marker before returning metadata

Reply:
That doesn't really answer my question. Where is this user metadata coming from? A DeleteMarker is caused by a DeleteObject API request. I don't see any ability to set user metadata. The AWS CLI does not allow metadata. So the question is, why do you need to apply user metadata?